build/curriculum.ts

```diff
@@ -0,0 +1,422 @@
+  const index = await buildIndex();
+  const i = index.findIndex((x) => x.slug === slug);
+  return prevNextFromIndex(i, index);
+}
+
+function breadPath(
+  url: string,
+  cur: CurriculumIndexEntry[]
+): DocParent[] | null {
+  for (const entry of cur) {
+    if (entry.url === url) {
+      return [{ uri: entry.url, title: entry.title }];
+    }
+    if (entry.children?.length) {
+      const found = breadPath(url, entry.children);
+      if (found) {
+        return [{ uri: entry.url, title: entry.title }, ...found];
+      }
+    }
+  }
+  return null;
+}
+
+async function buildParents(url: string): Promise<DocParent[]> {
+  const index = await buildCurriculumIndex(({ url, title }) => {
+    return { url, title };
+  });
+  const parents = breadPath(url, index);
+  if (parents) {
+    const { url, title } = index[0];
+    if (parents[0]?.uri !== url) {
+      return [{ uri: url, title }, ...parents];
+    }
+    return parents;
+  }
+
+  return [];
+}
+
+async function readCurriculumPage(
+  file: string,
+  options?: {
+    previousNext?: boolean;
+    forIndex?: boolean;
+  }
+): Promise<ReadCurriculum> {
+  const raw = await fs.readFile(file, "utf-8");
+  const { attributes, body: rawBody } = frontmatter<CurriculumFrontmatter>(raw);
+  const filename = file.replace(CURRICULUM_ROOT, "").replace(/^\/?/, "");
+  let title = rawBody.match(/^[\w\n]*#+(.*\n)/)[1]?.trim();
+  const body = rawBody.replace(/^[\w\n]*#+(.*\n)/, "");
+
+  const slug = fileToSlug(file);
+  const url = `/${DEFAULT_LOCALE}/${slug}/`;
+
+  let sidebar: CurriculumIndexEntry[];
+  let parents: DocParent[];
+
+  let modules: CurriculumIndexEntry[];
+  let prevNext: PrevNext;
+  if (!options?.forIndex) {
+    if (attributes.template === Template.landing) {
+      modules = (await buildCurriculumIndex())?.filter(
+        (x) => x.children?.length
+      );
+    } else if (attributes.template === Template.overview) {
+      modules = (await buildCurriculumIndex())?.find(
+        (x) => x.slug === slug
+      )?.children;
+    }
+    if (attributes.template === Template.module) {
+      prevNext = await buildPrevNextModule(slug);
+    } else if (attributes.template === Template.overview) {
+      prevNext = await buildPrevNextOverview(slug);
+    }
+
+    sidebar = await buildCurriculumSidebar();
+    parents = await buildParents(url);
+  } else {
+    title = title
+      .replace(/^\d+\s+/, "") // Strip number prefix.
+      .replace(/ modules$/, "") // Strip "modules" suffix.
+      .replace(/Extension \d+:/, ""); // Strip "Extension" prefix.
+  }
+
+  return {
+    meta: {
+      filename,
+      slug,
+      url,
+      title,
+      sidebar,
+      modules,
+      parents,
+      prevNext,
+      ...attributes,
+    },
+    body,
+  };
+}
+
+export async function findCurriculumPageBySlug(
+  slug: string
+): Promise<CurriculumData | null> {
+  let file = await slugToFile(slug);
+  if (!file) {
+    file = await slugToFile(`${slug}${slug ? "/" : ""}README`);
+  }
```

Comment:
Can we add a comment explaining the fallback?

Also, does this work on Windows if you use `"/"` instead of `path.join()` (which seems to ignore empty strings)?

Here's what I would suggest for better readability:

```suggestion
  const file = await slugToFile(slug) || await slugToFile(path.join(slug, "README"));
```